sdk/dotnet/Pulumi.Automation/Commands/LocalPulumiCmd.cs

```diff
@@ -131,7 +149,36 @@ internal class LocalPulumiCmd : IPulumiCmd
             proc.Start();
             proc.BeginOutputReadLine();
             proc.BeginErrorReadLine();
-            return await tcs.Task.ConfigureAwait(false);
+            try
+            {
+                return await tcs.Task.ConfigureAwait(false);
```

Comment:
@vipentti Do you think this may be an issue in the event that the cancellation token is triggered? Consider this scenario:

1. Process is running with event log watcher
2. Cancellation token is triggered
3. TaskCompletionSource is cancelled, triggering cleanup of event log file because `await tcs.Task` will return
4. Process is killed, but this may not happen instantaneously - it might still be trying to write to the log file

Reply:
That might indeed be the case, is the `Exited` event triggered also when the Process is killed? If so I think we could move the cleanup logic there after the `WaitForExit`

Reply:
yea I think the bottom of the `Exited` event might be the place for cleanup - I believe it is _always_ triggered but I can't find verification of that

Reply:
I did some testing trying to figure out in which cases the `Exited` event is called and in which cases it is not.
So far it looks like the `Exited` event **may not**  always be called, at least it was not in my contrived example.
For example when the `CancellationToken` is cancelled and a `System.Threading.Tasks.TaskCanceledException` is thrown,
the `Exited` event may not be called.

Basically what I did was add some extra `Console.Error.Writeline` calls before the `proc.Start()`, inside the `finally`-block,
inside `cancellationToken.Register` before `proc.Kill()` and inside `Exited` eventhandler after `proc.WaitForExit()`.

I then ran the following test code in Debug-mode to produce some debug output

```csharp
[Fact]
public async Task ProcessExitedEventOnCancellation()
{
    var program = PulumiFn.Create(() =>
    {
        return new Dictionary<string, object?>
        {
            ["exp_static"] = "foo",
        };
    });
    var projectName = "process_exited_test";
    var stackName = $"exited_on_cancel{GetTestSuffix()}";
    using var stack = await LocalWorkspace.CreateStackAsync(new InlineProgramArgs(projectName, stackName, program)
    {
        EnvironmentVariables = new Dictionary<string, string>()
        {
            ["PULUMI_CONFIG_PASSPHRASE"] = "test",
        }
    });

    try
    {
        var cts = new CancellationTokenSource();
        var previewTask = stack.PreviewAsync(cancellationToken: cts.Token);
        // Using cts.Cancel() produces similiar end results, with slightly different
        // results in the output
        cts.CancelAfter(0);
        await previewTask;
    }
    catch(Exception ex)
    {
        Console.Error.WriteLine("Uncaught exception: {0}", ex);
        throw;
    }
    finally
    {
        await stack.Workspace.RemoveStackAsync(stackName);
    }
}
```

Running this produces the following debug output (formatted for readability)

```yaml
# First two commands succeed and the order of the messages is the same
Start: Process 'version --non-interactive'
ExitedEvent: Process 'version --non-interactive' ExitCode=0
Finally: Process 'version --non-interactive' HasExited=True

Start: Process 'stack init exited_on_cancel191679 --non-interactive'
ExitedEvent: Process 'stack init exited_on_cancel191679 --non-interactive' ExitCode=0
Finally: Process 'stack init exited_on_cancel191679 --non-interactive' HasExited=True

# This call will fail due to the cancellation
Start: Process 'stack select exited_on_cancel191679 --non-interactive'
# NOTE 1: proc.HasExited is false here
# NOTE 2: No ExitedEvent for this command
# NOTE 3: The Register callback is called further below
Finally: Process 'stack select exited_on_cancel191679 --non-interactive' HasExited=False
# This is printed inside the test code above
Uncaught exception: System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at Pulumi.Automation.Commands.LocalPulumiCmd.RunAsync(IEnumerable`1 args, String workingDir, IDictionary`2 additionalEnv, Action`1 onStandardOutput, Action`1 onStandardError, Action`1 onEngineEvent, CancellationToken cancellationToken) in <path>/sdk/dotnet/Pulumi.Automation/Commands/LocalPulumiCmd.cs:line 161
   at Pulumi.Automation.WorkspaceStack.PreviewAsync(PreviewOptions options, CancellationToken cancellationToken) in <path>/sdk/dotnet/Pulumi.Automation/WorkspaceStack.cs:line 314
   <snip>

# This is the RemoveStackAsync call inside the test
Start: Process 'stack rm --yes exited_on_cancel191679 --non-interactive'

# Register callback gets called
Register: Killing process 'stack select exited_on_cancel191679 --non-interactive'
# proc.Kill() throws
Register: Killing process 'stack select exited_on_cancel191679 --non-interactive' error System.InvalidOperationException No process is associated with this object.
   at System.Diagnostics.Process.EnsureState(State state)
   at System.Diagnostics.Process.Kill()
   at Pulumi.Automation.Commands.LocalPulumiCmd.<>c__DisplayClass0_0.<RunAsync>b__2() in <path>/sdk/dotnet/Pulumi.Automation/Commands/LocalPulumiCmd.cs:line 114

# RemoveStackAsync succeeds
ExitedEvent: Process 'stack rm --yes exited_on_cancel191679 --non-interactive' ExitCode=0
Finally: Process 'stack rm --yes exited_on_cancel191679 --non-interactive' HasExited=True
```

At least based on this contrived example, running the cleanup in the `Exited` event handler may not always end up disposing the `EventLogWatcher`.

To ensure the process gets killed before removing the watcher and the file, when cancellation is requested, we could add the following inside the `finally`-block

```csharp
// The process may not have exited because cancellation was requested
// Ensure the process gets killed before cleaning up resources
if (!proc.HasExited)
{
    try { proc.Kill(); } catch { /* ignored */ }
    try { proc.WaitForExit(); } catch { /* ignored */ }
}
```

Reply:
It makes sense that if the process hasn't started yet than the `Exited` event wouldn't be triggered on cancellation.

Does it really need to be that complex though? Couldn't your finally block just be:
```csharp
if (!proc.HasExited)
    proc.WaitForExit();

// do cleanup
```

Reply:
Can we consider pro/con of taking a dependency here for example:

https://github.com/Tyrrrz/CliWrap 

This seems actively maintained and is a package trying to address the deficiencies of System.Diagnostics.Process API, in particular ability to safely async-wait and cancel.

Reply:
I think that is definitely worth considering and there is a good argument for that. But I think if we go that route it should be a separate PR after we get the log file watching working

Reply:
That sounds great! I can try drafting the PR if it sounds like a good idea. I'm also checking internally what is Pulumi's policy on dependencies, e.g. how conservative we want to be here.

Reply:
Or I guess it doesn't need to be after we get this working - if you think it would make managing process lifetime simpler. But I do think it should be a separate PR.

I think that I made a mistake not going with a library like that to manage the child process to begin with. But we could ask Mikhail or Justin their thoughts on taking the dependency - I think if it's fairly lightweight, actively maintained, and we're not exposing it publicly they won't mind at all.

Reply:
Yes that sounds wise, @mikhailshilkov any thoughts on adding CliWrap dependency. Justin is temporarily unavailable.

I am just looking at this code here and I remember struggling with something very similar circa 2013. The concurrency bugs around subprocess management times multiple Ones can take a toll. I do not know if the CliWrap ironed it all out but I like the idea of leaving it to OSS and solving this problem once and for all in there, if we can.

Reply:
Off the top of my head, I don't have any concerns about it.